examples/sugarscape-ig/ss_mesa/agents.py

```diff
@@ -0,0 +1,80 @@
+import math
+
+import mesa
+
+
+def get_distance(pos_1, pos_2):
+    """Get the distance between two point
+
+    Args:
+        pos_1, pos_2: Coordinate tuples for both points.
+    """
+    x1, y1 = pos_1
+    x2, y2 = pos_2
+    dx = x1 - x2
+    dy = y1 - y2
+    return math.sqrt(dx**2 + dy**2)
+
+
+class AntMesa(mesa.Agent):
+    def __init__(self, unique_id, model, moore=False, sugar=0, metabolism=0, vision=0):
+        super().__init__(unique_id, model)
+        self.moore = moore
+        self.sugar = sugar
+        self.metabolism = metabolism
+        self.vision = vision
+
+    def get_sugar(self, pos):
+        this_cell = self.model.grid.get_cell_list_contents([pos])
+        for agent in this_cell:
+            if type(agent) is Sugar:
+                return agent
+
+    def is_occupied(self, pos):
+        this_cell = self.model.grid.get_cell_list_contents([pos])
+        return any(isinstance(agent, AntMesa) for agent in this_cell)
+
+    def move(self):
+        # Get neighborhood within vision
+        neighbors = [
+            i
+            for i in self.model.grid.get_neighborhood(
+                self.pos, self.moore, False, radius=self.vision
+            )
+            if not self.is_occupied(i)
+        ]
+        neighbors.append(self.pos)
+        # Look for location with the most sugar
+        max_sugar = max(self.get_sugar(pos).amount for pos in neighbors)
+        candidates = [
+            pos for pos in neighbors if self.get_sugar(pos).amount == max_sugar
+        ]
+        # Narrow down to the nearest ones
+        min_dist = min(get_distance(self.pos, pos) for pos in candidates)
+        final_candidates = [
+            pos for pos in candidates if get_distance(self.pos, pos) == min_dist
+        ]
+        self.random.shuffle(final_candidates)
+        self.model.grid.move_agent(self, final_candidates[0])
+
+    def eat(self):
+        sugar_patch = self.get_sugar(self.pos)
+        self.sugar = self.sugar - self.metabolism + sugar_patch.amount
+        sugar_patch.amount = 0
+
+    def step(self):
+        self.move()
+        self.eat()
+        if self.sugar <= 0:
+            self.model.grid.remove_agent(self)
+            self.model.schedule.remove(self)
+
+
+class Sugar(mesa.Agent):
+    def __init__(self, unique_id, model, max_sugar):
+        super().__init__(unique_id, model)
+        self.amount = max_sugar
+        self.max_sugar = max_sugar
+
+    def step(self):
+        self.amount = self.max_sugar
```

Comment:
FLAME 2 folks did the growback only when unoccupied: https://github.com/FLAMEGPU/FLAMEGPU2-submodel-benchmark/blob/290702073492677f2fc9164419e9d5900d119e02/src/main.cu#L75.

Reply:
we can do that

Reply:
Yes, so that it is apple vs apple benchmark comparison. Even though this probably won't affect the runtime that much.